libbfddp/bfddp_extra.c

```diff
@@ -0,0 +1,858 @@
+/*
+ * BFD Data Plane library implementation.
+ *
+ * Copyright (C) 2020 Network Device Education Foundation, Inc. ("NetDEF")
+ *                    Rafael F. Zalamena
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the “Software”), to
+ * deal in the Software without restriction, including without limitation the
+ * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+ * sell copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include <sys/time.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "bfddp_extra.h"
+
+/*
+ * Callback handling.
+ */
+
+static int
+bfddp_session_new_dummy(__attribute__((unused)) struct bfd_session *bs,
+			__attribute__((unused)) void *arg)
+{
+	return 0;
+}
+
+static void
+bfddp_session_update_dummy(__attribute__((unused)) struct bfd_session *bs,
+			   __attribute__((unused)) void *arg)
+{
+}
+
+static void
+bfddp_session_state_change_dummy(
+	__attribute__((unused)) struct bfd_session *bs,
+	__attribute__((unused)) void *arg,
+	__attribute__((unused)) enum bfd_state_value ostate,
+	__attribute__((unused)) enum bfd_state_value nstate)
+{
+}
+
+/* Our selected BFD integration callbacks. */
+struct bfddp_callbacks bfddp_callbacks;
+
+void
+bfddp_initialize(struct bfddp_callbacks *bc)
+{
+	bfddp_callbacks = *bc;
+
+	if (bfddp_callbacks.bc_session_new == NULL)
+		bfddp_callbacks.bc_session_new = bfddp_session_new_dummy;
+	if (bfddp_callbacks.bc_session_update == NULL)
+		bfddp_callbacks.bc_session_update = bfddp_session_update_dummy;
+	if (bfddp_callbacks.bc_session_free == NULL)
+		bfddp_callbacks.bc_session_free = bfddp_session_update_dummy;
+	if (bfddp_callbacks.bc_state_change == NULL)
+		bfddp_callbacks.bc_state_change =
+			bfddp_session_state_change_dummy;
+
+#define CALLBACK_CHECK(cb)                                                     \
```

Comment:
Define the MACRO outside of the body of the function?

Reply:
It is only used here (and it is specialized)... Any reason to "export" it?

Reply:
I am not saying it's exported. I would keep it in the same translation unit, but having the definition of a MACRO function in the middle of the body affects readability.

Reply:
Sure, I'll move it out of the function.